src/python/pants/backend/typescript/dependency_inference/rules.py

```diff
@@ -0,0 +1,286 @@
+
+    source: TSSourceField
+    dependencies: TSDependenciesField
+
+
+class InferDepsTypeScriptSourceDependenciesRequest(InferDependenciesRequest):
+    """Target typescript_source."""
+
+    infer_from = TypeScriptSourceDependencyInferenceFieldSet
+
+
+@dataclass(frozen=True)
+class TypeScriptTestTargetFilepath:
+    filepath: str
+
+
+@dataclass(frozen=True)
+class InferredTypeScriptDependencies:
+    addresses: list[Address]
+
+
+@dataclass(frozen=True)
+class RequestWrapper:
+    request: Union[
+        InferDepsTypeScriptTestDependenciesRequest,
+        InferDepsTypeScriptTestsDependenciesRequest,
+        InferDepsTypeScriptSourcesDependenciesRequest,
+        InferDepsTypeScriptSourceDependenciesRequest,
+    ]
+
+
+def raise_dep_infer_failure(address: Address, path: str):
+    raise ValueError(
+        f"Failed to infer dependencies for target `{address}`.\n"
+        f"Could not process process inferred dependency on `{path}` expecting `{address}` to be present."
+    )
+
+
+@rule("Generic rule to infer dependencies of TypeScript targets.")
+async def infer_typescript_dependencies(
+    request_wrapper: RequestWrapper,
+) -> InferredTypeScriptDependencies:
+    """Infer dependencies for TypeScript targets."""
+    # depending on whether it's a target or target generator, get an appropriate field from the field set
+    # TODO: there must be a better way?
+    if hasattr(request_wrapper.request.field_set, "source"):
+        sources = await Get(
+            HydratedSources, HydrateSourcesRequest(request_wrapper.request.field_set.source)
+        )
+    else:
+        sources = await Get(
+            HydratedSources, HydrateSourcesRequest(request_wrapper.request.field_set.sources)
+        )
+
+    # TODO: we need to fetch all relevant tsconfig.json files that may contain the path mappings;
+    #  see https://www.typescriptlang.org/tsconfig#paths to learn more.
+    #  We also need jest.config.ts files for tests (and more config files as we add support for other test runners);
+    #  see https://jestjs.io/docs/configuration to learn more.
+    #  e.g. something like this:
+    # path_mapping_digest = await Get(Digest, PathGlobs(["**/*/tsconfig*.json"]))
+    # path_mapping_contents_result = await Get(DigestContents, Digest, path_mapping_digest)
+    # path_mapping = json.loads(path_mapping_contents_result[0].content.decode("utf-8"))
+    # logger.warning(path_mapping["compilerOptions"]["paths"])
+    # TODO: use the mapping to map the import paths
+
+    metadata = await _prepare_inference_metadata(request_wrapper.request.field_set.address)
+    import_strings = await Get(
+        NativeParsedJavascriptDependencies,
+        NativeDependenciesRequest(sources.snapshot.digest, metadata),
+    )
+    logger.warning((sources.snapshot.files[0], import_strings))
+
+    addresses = []
+
+    for path in import_strings.file_imports:
+        # check if an import is a directory (that should contain `index.ts` or `index.tsx`)
```

Comment:
This should be handled similarily to how `__init__.py` files are treated I believe, and then fed into the rust parser as a pattern via the metadata.

Reply:
Meaning a line such as
```ts
import * from "myMod"
```
should be inferred as `rootDir/myMod/index.ts`
because we ahead of time know all `index.ts` file locations, and have constructed patterns like:
```python
    {"myMod": "./myMod/index.ts"}
```